### PR TITLE
fix - firmware bugs for version 1.3.0

### PR DESCRIPTION
**Pull request by RobertLucian**
### Taken out of the documentation

- Fixed synchronization bugs which led to very small IO rates

    - Also caused values to overlap on other ports [#412](https://github.com/DexterInd/GrovePi/issues/412)
    - Caused I2C errors when functions were called with no delay between them [#409](https://github.com/DexterInd/GrovePi/issues/409)

- Small refactorization of the code to something slightly better
- Fixed the dust sensor [#408](https://github.com/DexterInd/GrovePi/issues/408)
- Mitigated abnormality caused by the Raspberry Pi not supporting clock stretching on the I2C [#411](https://github.com/DexterInd/GrovePi/issues/411)
- Debugged and sped up the rate of acquisition of the DHT sensor [#418](https://github.com/DexterInd/GrovePi/issues/418)
- Add IR receiver functionality for any remote control and fix inherent issues with the library [#416](https://github.com/DexterInd/GrovePi/issues/416)
- Add option to set the flow meter and the dust sensor on different ports other than D2 [#421](https://github.com/DexterInd/GrovePi/issues/421)
- Added documentation -> [Preview Docs while it's still hot](https://robertlucian.github.io/GrovePi)

### CODE REVIEW

**Comment by imrehg** [comment]
Hi, checked this with a DHT22 sensor (white Grove Temperature Humidity Sensor Pro v1.2), and while 1.2.7 firmware works, the new PR firmware doesn't give a proper reading.

**Comment by RobertLucian** [comment]
Does it only return values of 0.0 when using the DHT22 on the 1.3.0?

**Comment by RobertLucian** [comment]
@imrehg then your issue only confirms a suspicion I had regarding it. This problem has already been reported by someone on the forum, so that's great having you brought this up.

Just so you know, I placed an order for the DHT22 and I should be getting it next week (hopefully). Once I get it, I'll be able to debug it and fix it.

File: Firmware/Source/grovepi/extra_script.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Import('env')
```
Comment (CleoQc) [inline Firmware/Source/grovepi/extra_script.py]: Could you document this one a bit? With a name like "extra_script" and no info inside, it's hard to know what it's for.

File: Firmware/Source/grovepi/src/DHT.cpp [hunk 2]
```diff
@@ -154,12 +154,50 @@ boolean DHT::read(void) {
...
+      // temperature
+      buffer[0] = data[2];
+      if(scale)
+        buffer[0] = convertCtoF(buffer[0]);
+
+      // humidity
+      buffer[1] = data[0];
+      break;
+    case DHT22:
+      buffer[0] = buffer[1] = 0;
```
Comment (CleoQc) [inline Firmware/Source/grovepi/src/DHT.cpp]: oh is it where the DHT Pro always returns 0?

File: Firmware/Source/grovepi/src/IRremote.h [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+#define DECODE_DISH          0 // NOT WRITTEN
+#define SEND_DISH            1
+
+#define DECODE_SHARP         0 // NOT WRITTEN
+#define SEND_SHARP           1
+
+#define DECODE_DENON         1
+#define SEND_DENON           1
+
+#define DECODE_PRONTO        0 // This function doe not logically make sense
```
Comment (CleoQc) [inline Firmware/Source/grovepi/src/IRremote.h]: `does` , not `doe` :)

**Comment by deadprogram** [comment]
Hello there. Not sure if this is the correct place for this. I am testing the v1.3.0 firmware using our Go package Gobot's recentluy added support for GrovePi+. 

With a single analog device plugged into A1 is reading values as expected. however if I plug in a second device into either A0 or A2, than the reading from both analog devices is the same as the new device.

This is the code that communicates with the GrovePi here https://github.com/hybridgroup/gobot/blob/master/drivers/i2c/grovepi_driver.go#L194

To summarize: a single analog device seems to read fine, but 2 at the same time are both returning the same value.

**Comment by RobertLucian** [comment]
@deadprogram: When I had done my round of testing for the GrovePi with this firmware `v1.3.0`, I was able to poll the analog input ports simultaneously just fine, which means there has got to be something incompatible in your implementation.

I'm making 2 assumptions:
1. You implemented the API according to this version of the firmware and not the old one which had lots of bugs and which is not compatible with this one.
1. You're polling the ports in a single thread/process. The communication between the Pi and the GrovePi still isn't thread-safe and it won't be in the foreseeable future.

If both of my assumptions are correct, then I'm not sure what could be wrong with your implementation. Still, I could test the GrovePi again and see if polling multiple analog sensors continuously works - just to be 100% sure.

**Comment by RobertLucian** [comment]
Huh, it's interesting that the solution came from something that no one would have expected to help - an assumption. Anyway, that's fantastic news. Just for the record, we also added a mutex for one of our robots just because of that.

Let me know if you encounter any further problems with the firmware. If you want, you can send me emails from now on and I'll get back to you as soon as I can.

Thanks!

**Comment by RobertLucian** [comment]
@imrehg the DHT is now fixed. Sorry, it took me more than it should have, and that's due to slow customs and because I have so many things going on in parallel. Let me know if it works for you.

**Comment by imrehg** [comment]
Hi @RobertLucian how's this release going? Just wondering if there are any outstanding issues, or we can safely use it. Just flashed the 1.3.0 firmware from this PR as it stands, and so far it's working nicely.

**Comment by RobertLucian** [comment]
Hi @imrehg, in the past months, I haven't seen anyone having a problem with the beta except the ones we had already discussed here. No outstanding issues as far as we know.

Just to keep you updated, we still have to test this against the GrovePi Zero (which is the smaller brother of the GrovePi) and we still want to fully test it again against the GrovePi. At the moment we are a bit busy with other projects in the pipeline, such as the GiggleBot. 

And if for some reason, you encounter any issues with the firmware, please write them down and let me know about it.

**Comment by ductape** [comment]
Hi @RobertLucian I just wanted to let you know that I flashed this firmware and it's working fine for me. Specifically I tested the I2C changes with the lastest Raspbian and they work for me! I think this should close #409 (at least from my use case).

**Comment by RobertLucian** [comment]
Finally had the chance to test it - it works okay. 

There are still a few problems when continuously reading from the board, due to how the firmware was originally built, but they do not show up to the end-user, which is great. That's why I have added `try/except` blocks.
Also, I've seen problems with the four-digit display, so we'll probably have to come back to it. Not now, not when this firmware is better than the old one, the 1.2.7.

And I have switched the I2C driver and started using the DI SW I2C one which is a heck of a lot more reliable/stable. That should circumvent the inherent bugs with the HW I2C of the Raspberry Pi.
